src/suites/cts/robust_access.spec.ts

```diff
@@ -0,0 +1,364 @@
+      declaration: `${baseTypeName} data[]`,
+      length: 3,
+      std140Length: 0, // Unused
+      std430Length: 3,
+      zero: baseType.glslZero,
+      baseType,
+    };
+
+    // Vectors
+    for (let dimension = 2; dimension <= 4; dimension++) {
+      types[`${baseTypeName}_vector${dimension}`] = {
+        declaration: `${baseType.glslPrefix}vec${dimension} data`,
+        length: dimension,
+        std140Length: dimension,
+        std430Length: dimension,
+        zero: baseType.glslZero,
+        baseType,
+      };
+    }
+  }
+
+  // Matrices, there are only float matrics in GLSL.
+  for (const transposed of [true, false]) {
+    for (let firstDim = 2; firstDim <= 4; firstDim++) {
+      for (let secondDim = 2; secondDim <= 4; secondDim++) {
+        const majorDim = transposed ? secondDim : firstDim;
+        const minorDim = transposed ? firstDim : secondDim;
+
+        const std140SizePerMinorDim = 4;
+        const std430SizePerMinorDim = minorDim === 3 ? 4 : minorDim;
+
+        let typeName = `mat${firstDim}`;
+        if (firstDim !== secondDim) {
+          typeName += `x${secondDim}`;
+        }
+
+        types[(transposed ? 'transposed_' : '') + typeName] = {
+          declaration: (transposed ? 'layout(row_major) ' : '') + `${typeName} data`,
+          length: firstDim * secondDim,
+          std140Length: std140SizePerMinorDim * (majorDim - 1) + minorDim,
+          std430Length: std430SizePerMinorDim * (majorDim - 1) + minorDim,
+          zero: `vec${secondDim}(0.0f)`,
+          baseType: baseTypes['float'],
+        };
+      }
+    }
+  }
+
+  return types;
+})();
+
+g.test('bufferMemory', async t => {
+  await t.initGLSL();
+
+  const type = typeParams[t.params.type];
+  const baseType = type.baseType;
+
+  const indicesToTest = [
+    // Check exact bounds
+    `-1 * one`,
+    `${type.length} * one`,
+
+    // Check large offset
+    `-1000000 * one`,
+    `1000000 * one`,
+
+    // Check with max uint
+    `${kUintMax} * one`,
+    `-1 * ${kUintMax} * one`,
+
+    // Check with max int
+    `${kIntMax} * one`,
+    `-1 * ${kIntMax} * one`,
+  ];
+
+  const testSource = [];
+  let byteSize = 0;
+
+  // Declare the data that will be accessed to check robust access.
+  if (t.params.memory === 'uniform') {
+    testSource.push(`
+        layout(std140, set = 0, binding = 0) uniform TestData {
+          ${type.declaration};
+        };`);
+    byteSize = baseType.byteSize * type.std140Length;
+  } else {
+    testSource.push(`
+        layout(std430, set = 0, binding = 0) buffer TestData {
+          ${type.declaration};
+        };`);
+    byteSize = baseType.byteSize * type.std430Length;
+  }
+
+  // Build the test function that will do the tests.
+  testSource.push(`
+    uint runTest() {
+  `);
+
+  for (const index of indicesToTest) {
+    if (t.params.access === 'read') {
+      testSource.push(`
+          if(data[${index}] != ${type.zero}) {
+            return __LINE__;
```

Comment:
nit: Might be slightly confusing because the __LINE__ doesn't really map to anything interesting since the GLSL is generated. How about 0x0BAD or something just to make it ctrl-f-able?

Reply:
This one is supposed to help debug which test failed. I'd like to make a nicer shader test harness at some point that reports errors with a string.

Reply:
Makes sense